File: dsa/5.binary_search/441_Arranging_Coins.java
package binary_search;

class Solution {
    public int arrangeCoins(int n) {
        int left = 1;
        int right = n;

        while (left <= right){
            int mid = left + (right-left)/2;
            if ((long) (mid + 1)*mid/2 == (long)n){
                return mid;
            }
            else if (((long) (mid + 1)*mid/2 > (long)n) & ((long) (mid + 1)*mid/2 - mid <= (long)n)){
                return mid - 1;
            }
            else if((long) (mid + 1)*mid/2 > (long)n){
                right = mid - 1;
            }
            else {
                left = mid + 1;
            }
        }
        return -1;
    }
}

File: dsa/5.binary_search/441_Arranging_Coins.py
class Solution:
    def arrangeCoins(self, n: int) -> int:
        left: int = 0
        right: int = n
        while left <= right:
            mid: int = int((right - left) / 2) + left
            cur_sum_val: int = self.sum(mid)
            next_sum_val: int = self.sum(mid + 1)
            if n >= cur_sum_val and n < next_sum_val:
                return mid
            elif n < cur_sum_val:
                right = mid - 1
            else:
                left = mid + 1
        
        return -1 
    
    def sum(self, n: int) -> int:
        return (n + 1) * n / 2